File: backend/app/schemas/kalem.py
from decimal import Decimal
from enum import Enum
from typing import Annotated, Any
from pydantic import BaseModel, BeforeValidator, field_validator


class BeyannameBolumuEnum(str, Enum):
    ZARAR_OLSA_DAHI = "zarar_olsa_dahi"
    KAZANC_VARSA = "kazanc_varsa"
    GECMIS_YIL_ZARARI = "gecmis_yil_zarari"
    ILAVE = "ilave"
    HESAPLANAN_KV_INDIRIMI = "hesaplanan_kv_indirimi"


class YiakVEtkisiEnum(str, Enum):
    DUSULUR = "dusulur"
    DUSULMEZ = "dusulmez"
    TARTISMALI = "tartismali"


class DurumEnum(str, Enum):
    AKTIF = "aktif"
    GECIS = "gecis"
    MULGA = "mulga"


class VeriGirisiTipiEnum(str, Enum):
    PARA = "para"
    TARIH = "tarih"
    SECENEK = "secenek"
    EVET_HAYIR = "evet_hayir"
    METIN = "metin"
    SAYI = "sayi"
    BOLUM = "bolum"
    MATRIS = "matris"
    UZLASTIRMA = "uzlastirma"


class BelgeKategorisiEnum(str, Enum):
    ZORUNLU = "zorunlu"
    DESTEKLEYICI = "destekleyici"


class BeyannamKodu(BaseModel):
    donem: int
    kod: int
    aciklama: str | None = None


class YururlukAraligi(BaseModel):
    baslangic: str
    bitis: str | None = None


class KapiSorusu(BaseModel):
    id: str
    soru: str
    tip: str
    zorunlu_cevap: str | None = None
    aciklama: str | None = None


class WizardAgaci(BaseModel):
    tetikleyici_soru: str
    info_modal: str | None = None
    kapi_sorulari: list[KapiSorusu] = []


def _secenekler_normalize(v: Any) -> Any:
    """str → {deger, etiket} dönüşümü — hem plain string hem dict kabul eder."""
    if isinstance(v, list):
        return [{"deger": x, "etiket": x} if isinstance(x, str) else x for x in v]
    return v


class SecebekItem(BaseModel):
    """Zengin seçenek: {deger, etiket}. YAML'da hem plain string hem dict kabul edilir."""
    deger: str
    etiket: str


class VeriGirisiAlani(BaseModel):
    id: str
    etiket: str
    tip: VeriGirisiTipiEnum
    zorunlu: bool = False
    varsayilan: str | int | float | bool | None = None
    secenekler: Annotated[list[SecebekItem] | None, BeforeValidator(_secenekler_normalize)] = None
    yardim: str | None = None
    satirlar: list[dict[str, Any]] | None = None   # matris rows: [{id, etiket}]
    sutunlar: list[dict[str, Any]] | None = None   # matris cols: [{id, etiket, zorunlu?}]
    kaynak_alan: str | None = None                  # uzlastirma: watched form field id


class Validasyon(BaseModel):
    kural: str
    hata: str


class HesaplamaSablonu(BaseModel):
    veri_girisi_alanlari: list[VeriGirisiAlani] = []
    formuller: dict[str, str] = {}
    sonuc_alan: str = ""
    validasyonlar: list[Validasyon] = []


class BelgeListesiKalem(BaseModel):
    kategori: BelgeKategorisiEnum
    no: int
    baslik: str
    detay: str | None = None
    temin_yeri: str | None = None


class KChecklistItem(BaseModel):
    id: str
    soru: str
    referans: str | None = None


class KalemSchema(BaseModel):
    ic_kod: str
    baslik: str
    ust_kalem: str | None = None
    beyanname_kodlari: list[BeyannamKodu] = []
    dahili_ref: str | None = None  # XML'de karşılığı olmayan özel iç referans (ör. "IKV-YTB")
    mevzuat_dayanagi: list[str] = []
    beyanname_bolumu: BeyannameBolumuEnum
    yiakv_etkisi: YiakVEtkisiEnum
    durum: DurumEnum = DurumEnum.AKTIF
    yururluk_araligi: YururlukAraligi | None = None
    ana_kategori: str | None = None
    wizard_agaci: WizardAgaci | None = None
    parametreler: dict[str, Any] = {}
    coklu_instance: bool = False  # aynı kalemden birden fazla eklenebilir (ör. YTB)
    hesaplama_sablonu: HesaplamaSablonu = HesaplamaSablonu()
    denetci_notlari: str | None = None
    muhasebe_kayitlari: str | None = None
    belge_listesi: list[BelgeListesiKalem] = []
    k_checklist: list[KChecklistItem] = []
    hesaplayici: str | None = None  # karmaşık Python hesaplayıcı referansı

    @field_validator("ic_kod")
    @classmethod
    def ic_kod_bos_olamaz(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("ic_kod boş olamaz")
        return v

    def get_beyanname_kodu(self, donem_yili: int) -> int | None:
        """Dönem yılına göre GİB beyanname kodunu döner."""
        for bk in self.beyanname_kodlari:
            if bk.donem == donem_yili:
                return bk.kod
        # En yakın önceki dönemi bul
        onceki = [bk for bk in self.beyanname_kodlari if bk.donem <= donem_yili]
        if onceki:
            return max(onceki, key=lambda x: x.donem).kod
        return None

File: frontend/src/api/kalem.ts
import { useQuery, useMutation } from '@tanstack/react-query'
import { apiClient } from './client'

export interface AnaKategoriPublic {
  id: number
  kod: string
  soru: string
  etiket: string
  bilgi: string | null
  grup: 'zarar_olsa_dahi' | 'kazanc_varsa' | 'gecmis_yil_zarari' | 'ilave' | 'hesaplanan_kv_indirimi'
  beyanname_kodlari: number[] | null
  sira: number
}

export function useAnaKategoriler() {
  return useQuery({
    queryKey: ['katalog', 'ana-kategoriler'],
    queryFn: () =>
      apiClient.get('/katalog/ana-kategoriler').then((r) => r.data as AnaKategoriPublic[]),
    staleTime: 5 * 60 * 1000, // 5 min cache
  })
}

export interface KatalogKalem {
  ic_kod: string
  baslik: string
  ana_kategori: string
  beyanname_bolumu: string
  yiakv_etkisi: string
  durum: string
  coklu_instance?: boolean
  dahili_ref?: string | null
  beyanname_kodlari?: { donem: number; kod: number }[]
}

export interface MatrisSatirBilesen {
  satir: string
  isaret: 1 | -1
}

export interface MatrisSatir {
  id: string
  etiket: string
  tip?: 'satir' | 'bolum' | 'ara_toplam' | 'dagilim_secim'
  bilesenler?: MatrisSatirBilesen[]
  grup?: string  // e.g. 'faaliyet_gideri'
}

export interface MatrisSutun {
  id: string
  etiket: string
  zorunlu?: boolean
  renk?: string
  hesapli?: boolean  // auto-computed column, no user input
}

export interface VeriGirisiAlani {
  id: string
  etiket: string
  tip: 'para' | 'tarih' | 'secenek' | 'evet_hayir' | 'metin' | 'sayi' | 'matris' | 'bolum' | 'uzlastirma'
  zorunlu: boolean
  secenekler?: { deger: string; etiket: string }[]
  yardim?: string
  varsayilan?: unknown
  satirlar?: MatrisSatir[]
  sutunlar?: MatrisSutun[]
  kaynak_alan?: string  // uzlastirma: which form field holds the dönem karı
}

export interface KChecklistMaddesi {
  id: string
  soru: string
  referans?: string  // backend field name
}

export interface BelgeMaddesi {
  no: number         // backend sends 'no' as integer
  baslik: string
  kategori: 'zorunlu' | 'destekleyici'
  detay?: string
  temin_yeri?: string
}

export interface HesaplamaSablonu {
  veri_girisi_alanlari: VeriGirisiAlani[]
}

export interface KalemSchema {
  ic_kod: string
  baslik: string
  kisa_aciklama?: string
  coklu_instance?: boolean
  beyanname_kodlari?: { donem: number; kod: number }[]
  mevzuat_dayanagi?: string[]
  wizard_agaci?: {
    tetikleyici_soru?: string
    info_modal?: string
  }
  hesaplama_sablonu: HesaplamaSablonu
  k_checklist: KChecklistMaddesi[]
  belge_listesi: BelgeMaddesi[]
}

export interface HesapSonucu {
  ic_kod: string
  istisna_tutari: number
  ara_sonuclar: Record<string, number>
  hatalar: string[]
  uyarilar: string[]
  aciklama: string
}

export interface ChecklistDurum {
  [maddeId: string]: 'uygun' | 'eksik' | 'risk'
}

export interface BelgeDurum {
  [belgeNo: string]: {
    durum: 'uygun' | 'eksik'
    not: string
  }
}

export function useKatalogKalemler() {
  return useQuery<KatalogKalem[]>({
    queryKey: ['katalog', 'kalemler'],
    queryFn: async () => {
      const { data } = await apiClient.get<KatalogKalem[]>('/katalog/kalemler')
      return data
    },
    staleTime: 1000 * 60 * 60, // 1 hour — catalog rarely changes
  })
}

export interface KalemVeriResponse {
  girdi_verileri: Record<string, unknown> | null
  istisna_tutari: number | null
  ara_sonuclar: Record<string, number> | null
  k_checklist_durumu: ChecklistDurum | null
  belge_durumu: BelgeDurum | null
}

export function useKalemVeri(calismaId: string | undefined, icKod: string | undefined) {
  return useQuery<KalemVeriResponse>({
    queryKey: ['kalem-veri', calismaId, icKod],
    queryFn: async () => {
      const { data } = await apiClient.get<KalemVeriResponse>(
        `/calisma/${calismaId}/kalem/${icKod}/veri`
      )
      return data
    },
    enabled: !!calismaId && !!icKod,
  })
}

export function useKalemSchema(icKod: string | undefined) {
  return useQuery<KalemSchema>({
    queryKey: ['kalem-schema', icKod],
    queryFn: async () => {
      const { data } = await apiClient.get<KalemSchema>(`/katalog/kalemler/${icKod}`)
      return data
    },
    enabled: !!icKod,
  })
}

export function useHesapla(calismaId: string | undefined, icKod: string | undefined) {
  return useMutation<HesapSonucu, Error, Record<string, unknown>>({
    mutationFn: async (girdiVerileri) => {
      if (!calismaId) throw new Error('calismaId is required')
      if (!icKod) throw new Error('icKod is required')
      const { data } = await apiClient.post<HesapSonucu>(
        `/calisma/${calismaId}/kalem/${icKod}/hesapla`,
        { girdi_verileri: girdiVerileri }
      )
      return data
    },
  })
}

export function useSaveVeri(calismaId: string | undefined, icKod: string | undefined) {
  return useMutation<void, Error, Record<string, unknown>>({
    mutationFn: async (girdiVerileri) => {
      if (!calismaId) throw new Error('calismaId is required')
      if (!icKod) throw new Error('icKod is required')
      await apiClient.put(
        `/calisma/${calismaId}/kalem/${icKod}/veri`,
        { girdi_verileri: girdiVerileri }
      )
    },
  })
}

export function useUpdateChecklist(calismaId: string | undefined, icKod: string | undefined) {
  return useMutation<void, Error, ChecklistDurum>({
    mutationFn: async (durum) => {
      if (!calismaId) throw new Error('calismaId is required')
      if (!icKod) throw new Error('icKod is required')
      await apiClient.put(
        `/calisma/${calismaId}/kalem/${icKod}/checklist`,
        { durum }
      )
    },
  })
}

export function useUpdateBelgeler(calismaId: string | undefined, icKod: string | undefined) {
  return useMutation<void, Error, BelgeDurum>({
    mutationFn: async (durum) => {
      if (!calismaId) throw new Error('calismaId is required')
      if (!icKod) throw new Error('icKod is required')
      await apiClient.put(
        `/calisma/${calismaId}/kalem/${icKod}/belgeler`,
        { durum }
      )
    },
  })
}
